srk/src/termination.ml

```diff
@@ -0,0 +1,429 @@
+  if n = 1 then
+    let s = mk_const srk (mk_symbol srk ~name:name `TyReal) in
+      [s]
+  else
+    let s = mk_const srk (mk_symbol srk ~name:name `TyReal) in
+      s :: create_const_symbols srk name (n-1)
+
+let build_formula srk lambdas mat n_rows n_cols op =
+  let m = Mat.transpose mat in
+  let formulas = ref (mk_true srk) in
+  for i = 0 to n_cols-1 do
+    let row_i = Mat.row i m in
+    let lhs = (mk_zero srk) in
+    let rlhs = ref lhs in
+    for j = 0 to n_rows-1 do
+      let ith_ele = Vec.coeff j row_i in
+      let term_v = mk_real srk ith_ele in
+      let prod = mk_mul srk [ List.nth lambdas j; term_v] in
+      let value = mk_add srk [ prod; !rlhs] in
+      rlhs := value
+    done;
+    let lhs_exp = !rlhs in
+    let formula = match op with
+    | "eq_0" -> mk_eq srk lhs_exp (mk_zero srk)
+    | "leq_0" -> mk_leq srk lhs_exp (mk_zero srk)
+    | "lt_0" -> mk_lt srk lhs_exp (mk_zero srk)
+    | _ -> assert false
+    in
+    formulas := mk_and srk [formula; !formulas]
+  done;
+  !formulas
+
+let build_lambda_mat_prod_terms srk lambdas mat n_rows n_cols =  
+  let m = Mat.transpose mat in
+  let r = ref [] in
+  for i = 0 to n_cols-1 do
+    let row_i = Mat.row i m in
+    let lhs = (mk_zero srk) in
+    let rlhs = ref lhs in
+    for j = 0 to n_rows-1 do
+      let ith_ele = Vec.coeff j row_i in
+      let term_v = mk_real srk ith_ele in
+      let prod = mk_mul srk [ List.nth lambdas j; term_v] in
+      let value = mk_add srk [ prod; !rlhs] in
+      rlhs := value
+    done;
+    r := List.concat [!r; [!rlhs]]
+  done;
+  !r
+
+let build_gt_zero srk lambdas =
+  List.fold_left
+    (fun f lambda ->
+      mk_and srk [f; mk_leq srk (mk_zero srk) lambda]
+    )
+    (mk_true srk)
+    lambdas
+
+let rec build_subtracted_lambdas srk lambdas1 lambdas2 =
+  match lambdas1, lambdas2 with
+  | h1 :: t1, h2 :: t2 -> 
+    mk_sub srk h1 h2 :: build_subtracted_lambdas srk t1 t2
+  | [], [] -> []
+  | _, _ -> assert false
+
+let get_constants_vec cs constraints =
+  BatList.fold_left
+    (fun (existing_rows, i) eq -> 
+        match eq with
+        | `LeqZero t -> 
+          begin
+            let row = Vec.coeff CoordinateSystem.const_id t in
+            let neg = QQ.negate row in
+            let m =  Vec.add_term neg i existing_rows in
+            (m, i+1)
+          end
+        | `LtZero t -> Log.error "there should not be strict inequalities"; assert false
+        | `EqZero t -> 
+          begin
+            let row = Vec.coeff CoordinateSystem.const_id t in
+            let neg = QQ.negate row in
+            let m =  Vec.add_term neg i existing_rows in
+            let row2 = Vec.coeff CoordinateSystem.const_id t in
+            let m =  Vec.add_term row2 (i+1) m in
+            (m, i+2)
+          end
+    )
+    (Vec.zero, 0)
+    constraints
+  
+let print_formula_to_stdout srk descrip f =
+  logf descrip;
+  let s2 = (Printf.sprintf "\n%s\n" (Formula.show srk f)) in
+  logf "%s" s2;
```

Comment:
logf "@\n%a@\n" (Formula.pp srk) f

See the OCaml Format module for more information about how to use logf & errorf.